### PR TITLE
Broadcasting for SLN for CPU and CUDA

### PR DESCRIPTION
**Pull request by khspear**
### Description
Enhanced SkipLayerNorm by implementing broadcasting for both CPU and CUDA

### Motivation and Context
The input and skip tensors no longer have to be the same size which means that it can accept data where the skip shape can be the same size as the input shape, have a shape of {1, sequence_length, hidden_size}, or {sequence_length, hidden_size}.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -39,5 +39,82 @@
     "cpplint.filters": [
         "-build/include_subdir",
         "-runtime/references"
-    ]
+    ],
+    "cmake.sourceDirectory": "/home/t-khspear/onnxruntime/cmake/external/libprotobuf-mutator",
```
Comment (tianleiwu) [inline .vscode/settings.json:43]: revert this file

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 2]
```diff
@@ -45,17 +45,16 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
...
+  size_t skip_dims_size = skip_dims.size();
+
   if (input_dims_size != 3 && input_dims_size != 2) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                            "input is expected to have 3 or 2 dimensions, got ", input_dims_size);
   }
 
   int hidden_size = static_cast<int>(input_dims[input_dims_size - 1]);
 
-  if (input->Shape() != skip->Shape()) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:55]: Shall check that input and skip has same shape, or skip_dims[0] == 1 and can be broadcasted to input shape...

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 3]
```diff
@@ -105,20 +104,27 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
   // of the input and skip tensors
   T* skip_input_bias_add_output_data = skip_input_bias_add_output != nullptr ? skip_input_bias_add_output->MutableData<T>() : nullptr;
 
+  const bool skip_broadcasted = (nullptr != skip && skip->Shape()[0] == 1) ? true : false;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:107]: `nullptr != skip` is not needed since skip is not optional.

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc [hunk 4]
```diff
@@ -60,17 +60,14 @@ Status SkipLayerNorm<T, Simplified>::ComputeInternal(OpKernelContext* ctx) const
   // of the input and skip tensors
   Tensor* skip_input_bias_add_output = ctx->Output(3, input->Shape());
 
-  if (input->Shape() != skip->Shape()) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc:63]: Need update here same as in cpu.

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu [hunk 5]
```diff
@@ -82,18 +82,23 @@ template <typename T, unsigned TPB, bool Simplified>
...
   cub::KeyValuePair<T, T> thread_data(0, 0);
 
+
   for (int i = threadIdx.x; i < ld; i += TPB) {
     const int idx = offset + i;
 
     const T val = (bias == nullptr) ? input[idx] + skip[idx] : input[idx] + skip[idx] + bias[i];
+
+    if (skip_broadcasted){
+      const T val = (bias == nullptr) ? input[idx] + skip[offset % skip_size] : input[idx] + skip[offset % skip_size] + bias[i];
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu:100]: This val has no effect.   Update line 96 instead.

File: onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc [hunk 6]
```diff
@@ -27,18 +27,32 @@ static void RunTest(
...
     input_dims = {batch_size * sequence_length, hidden_size};
   }
-  std::vector<int64_t> skip_dims = input_dims;
+
+  if(batch_size_1 == true){
+    skip_dims = {1, sequence_length, hidden_size};
+  }
+
+  if(no_batch_size){
+    skip_dims = {sequence_length, hidden_size};
```
Comment (tianleiwu) [inline onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc:53]: Do we support this format? It is not in the comment of line 35.

File: onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc [hunk 7]
```diff
@@ -27,18 +27,32 @@ static void RunTest(
     bool no_beta = false,
     bool simplified = false,
     bool use_token_count = false,
-    bool strict = false) {
+    bool strict = false,
+    bool batch_size_1 = false,
```
Comment (tianleiwu) [inline onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc:31]: Rename batch_size_1 to broadcast_skip

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 8]
```diff
@@ -105,20 +104,27 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
...
+  int skip_size = static_cast<int>(skip_dims[skip_dims_size - 1] * skip_dims[skip_dims_size - 2]);
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:113]: We can use one parameter to replace `skip_broadcasted` and `skip_size`:
```
skip_stride = skip->Shape().Size();   # total elements in skip
```

Then use `offset % skip_stride` instead of `(skip_broadcasted ? (offset % skip_size) : offset)`

**Review by tianleiwu — CHANGES_REQUESTED** [review]
Need update the operator spec:
https://github.com/microsoft/onnxruntime/blob/efeb6672d6492d53882ab265d4016aafae9e2233/onnxruntime/core/graph/contrib_ops/bert_defs.cc#L967

"Azure Pipelines / Windows GPU CI Pipeline (kernelDocumentation build_x64_RelWithDebInfo)"  will fail after the spec is changed. You can copy the artifacts of failed build and update the corresponding document files.

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc (hunk 8 above)
Comment (khspear) [reply onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:129]: So offset % skip_size (after the edit to skip_size) would account for if the input and skip tensor were the same size and if the skip tensor has a batch size of one?

File: onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc (hunk 6 above)
Comment (khspear) [reply onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc:53]: My apologies, I added some new edits to support that format.

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu (hunk 5 above)
Comment (khspear) [reply onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu:100]: There is nothing at line 96; do you mean line 97 or line 95? And what do you mean by that val has no effect?

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc (hunk 8 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:131]: Right. The maximum offset is number of elements in input tensor minus one.
For the case that input and skip having same shape, offset < input_size == skip_size, so offset % skip_size == offset.
You can see that the `offset % skip_size` have same value as `skip_broadcasted ? (offset % skip_size) : offset`

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 9]
```diff
@@ -45,18 +45,40 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
 
   const auto& input_dims = input->Shape().GetDims();
   size_t input_dims_size = input_dims.size();
+  const auto& skip_dims = skip->Shape().GetDims();
+  size_t skip_dims_size = skip_dims.size();
+  int skip_sequence_length = static_cast<int>(skip_dims[skip_dims_size - 2]);
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:50]: Need to check skip_dims_size == 3 before this line. Otherwise skip_dims[skip_dims_size - 2] might throw exeception when skip_dims_size < 2.

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 10]
```diff
@@ -45,18 +45,40 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
...
+
+  if (input->Shape() != skip->Shape() && skip_dims[0] != 1 && skip_dims[skip_dims_size - 2] != skip_sequence_length) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:52]: skip_dims[skip_dims_size - 2] != skip_sequence_length is not needed here since `int skip_sequence_length = static_cast<int>(skip_dims[skip_dims_size - 2]);`

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:55]: Based on your recent code, I assume that we support the following cases:
(1)   input and skip: (batch_size, sequence_length, hidden_size)
(2)  input and skip: (batch_size * sequence_length, hidden_size)
(3)   input: (batch_size, sequence_length, hidden_size), skip: (1, sequence_length, hidden_size)
(4)  input: (batch_size, sequence_length, hidden_size), skip: (sequence_length, hidden_size)

Is my understanding correct?

Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:55]: Yes, this PR adds (3) and (4)

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 11]
```diff
@@ -45,18 +45,42 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
...
+                           "skip is expected to have same shape as input or a batch size of 1 or no batch size");
+  }
+
   if (input_dims_size != 3 && input_dims_size != 2) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                            "input is expected to have 3 or 2 dimensions, got ", input_dims_size);
   }
 
+
+  if (skip_sequence_length != input_dims[1] && skip_dims[skip_dims_size - 2] != 1) {
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:69]: skip_seq_len doesn't need to be 1, it's the sequence length:
the added skip shape: (1, sequence_length, hidden_size), (sequence_length, hidden_size)

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 12]
```diff
@@ -45,18 +45,42 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
...
+                           "sequence length needs to be 1 or same as input");
+  }
+
   int hidden_size = static_cast<int>(input_dims[input_dims_size - 1]);
 
-  if (input->Shape() != skip->Shape()) {
+  if (skip_dims[skip_dims_size - 1] != hidden_size && skip_dims[skip_dims_size - 1] != 1) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
-                           "skip is expected to have same shape as input");
+                           "hidden size needs to be 1 or same as input");
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:78]: hidden size doesn't need to be 1

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc [hunk 13]
```diff
@@ -60,24 +60,41 @@ Status SkipLayerNorm<T, Simplified>::ComputeInternal(OpKernelContext* ctx) const
...
-  size_t input_dims_size = input_dims.size();
   if (input_dims_size != 3 && input_dims_size != 2) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                            "input is expected to have 3 or 2 dimensions, got ", input_dims_size);
   }
 
+
+  if (skip_sequence_length != input_dims[1] && skip_dims[skip_dims_size - 2] != 1) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "sequence length needs to be 1 or same as input");
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc:88]: ditto

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc [hunk 14]
```diff
@@ -60,24 +60,41 @@ Status SkipLayerNorm<T, Simplified>::ComputeInternal(OpKernelContext* ctx) const
...
+  }
+
   int hidden_size = static_cast<int>(input_dims[input_dims_size - 1]);
 
+  if (skip_dims[skip_dims_size - 1] != hidden_size && skip_dims[skip_dims_size - 1] != 1) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "hidden size needs to be 1 or same as input");
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc:95]: ditto

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc [hunk 15]
```diff
@@ -114,6 +131,9 @@ Status SkipLayerNorm<T, Simplified>::ComputeInternal(OpKernelContext* ctx) const
     }
   }
 
+  const bool skip_broadcasted = (skip->Shape()[0] == 1 || skip->Shape()[skip_dims_size-2] == skip_sequence_length) ? true : false;
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc:134]: skip->Shape()[skip_dims_size-2] == skip_sequence_length is always true.
I think you want to compare skip_seq_len with input_seq_len here.

const bool skip_broadcasted = skip_dims[0] == 1 || skip_dims_size  == 2

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu [hunk 16]
```diff
@@ -82,18 +82,23 @@ template <typename T, unsigned TPB, bool Simplified>
...
   // reduce x and x^2
   cub::KeyValuePair<T, T> thread_data(0, 0);
 
+
   for (int i = threadIdx.x; i < ld; i += TPB) {
     const int idx = offset + i;
 
     const T val = (bias == nullptr) ? input[idx] + skip[idx] : input[idx] + skip[idx] + bias[i];
+
+    if (skip_broadcasted){
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu:99]: you can do something like
const T skip_data = skip_broadcasted ? skip[idx % skip_size] : skip[idx]
and then use skip_data

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu [hunk 17]
```diff
@@ -82,18 +82,23 @@ template <typename T, unsigned TPB, bool Simplified>
...
+      val = (bias == nullptr) ? input[idx] + skip[offset % skip_size] : input[idx] + skip[offset % skip_size] + bias[i];
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu:100]: idx % skip_size ?

**Comment by wangyems** [comment]
you can use the lintrunner to format the code
https://github.com/microsoft/onnxruntime/actions/runs/5427073104/jobs/9869920517?pr=16510

**Comment by tianleiwu** [comment]
Some build pipeline failed. Please fix required pipelines first.
For example, to fix python format pipeline:
https://github.com/microsoft/onnxruntime/blob/main/docs/Coding_Conventions_and_Standards.md#linting

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 18]
```diff
@@ -105,20 +117,26 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
   // of the input and skip tensors
   T* skip_input_bias_add_output_data = skip_input_bias_add_output != nullptr ? skip_input_bias_add_output->MutableData<T>() : nullptr;
 
+  skip_size = skip->Shape().Size();
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:120]: Need add data type like `const auto`

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 19]
```diff
@@ -45,18 +45,30 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
...
+  if (skip_dims_size != 3 && skip_dims_size != 2) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
-                           "input is expected to have 3 or 2 dimensions, got ", input_dims_size);
+                           "skip is expected to have 3 or 2 dimensions, got ", skip_dims_size);
   }
+  
+  int skip_sequence_length = static_cast<int>(skip_dims[skip_dims_size - 2]);
 
-  int hidden_size = static_cast<int>(input_dims[input_dims_size - 1]);
+  if (input->Shape() != skip->Shape() && skip_dims[0] != 1 && skip_dims[0] != skip_sequence_length) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:58]: This check is not enough. We need add a condition that last two dimensions are same for input and skip like the following:
```
if (input->Shape() != skip->Shape() && input_dim_size != 3 && (skip_dims[0] != 1 || skip_dim_size != 2) && skip_dims[skip_dim_size-1] != input_dims[input_dim_size-1] && skip_dims[skip_dim_size-2] != input_dims[input_dim_size-2]) {
```

In this way, we can cover the following cases:
(1) input and skip: (batch_size, sequence_length, hidden_size)
(2) input and skip: (batch_size * sequence_length, hidden_size)
(3) input: (batch_size, sequence_length, hidden_size), skip: (1, sequence_length, hidden_size)
(4) input: (batch_size, sequence_length, hidden_size), skip: (sequence_length, hidden_size)

Comment (khspear) [reply onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:58]: Could these work?:

if (skip_dims[skip_dims_size - 2] != input_dims[input_dims_size - 2]) {
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                           "sequence length has to be the same for skip and input");
  }

  if (skip_dims[skip_dims_size - 1] != input_dims[input_dims_size - 1]) {
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                           "last dimension of skip needs to be same as input");
  }
  if (input->Shape() != skip->Shape() && skip_dims[0] != 1 && skip_dims[0] != skip_sequence_length && input_dims_size != 3) {
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                           "skip is expected to have same shape as input or, a batch size of 1 or no batch size when input has 3 dimensions");
  }

Comment (khspear) [reply onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:58]: Actually, for that last check, this looks better:

if ((input->Shape() != skip->Shape()) && ((skip_dims[0] != 1 || skip_dims_size != 2) && input_dims_size != 3)) {
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                           "skip is expected to have same shape as input or, a batch size of 1 or no batch size when input has 3 dimensions");
  }

Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:58]: That looks good to me.

Comment (khspear) [reply onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:58]: Okay, I see what you're saying. So, combine the first two checks into one and keep the last one as is?

Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:58]: Right

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 20]
```diff
@@ -45,18 +45,34 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
...
+  const auto& skip_dims = skip->Shape().GetDims();
+  size_t skip_dims_size = skip_dims.size();
+
+  if (skip_dims_size != 3 && skip_dims_size != 2) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
-                           "input is expected to have 3 or 2 dimensions, got ", input_dims_size);
+                           "skip is expected to have 3 or 2 dimensions, got ", skip_dims_size);
+  }
+  
+  int skip_sequence_length = static_cast<int>(skip_dims[skip_dims_size - 2]);
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:56]: skip_sequence_length is not used. It can be removed.

File: onnxruntime/contrib_ops/cpu/skip_layer_norm.cc [hunk 21]
```diff
@@ -45,18 +45,34 @@ Status SkipLayerNorm<T>::Compute(OpKernelContext* p_ctx) const {
...
+
+  if ((input->Shape() != skip->Shape()) && ((skip_dims[0] != 1 || skip_dims_size != 2) && input_dims_size != 3)) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                            "skip is expected to have same shape as input or, a batch size of 1 or no batch size when input has 3 dimensions");
   }
 
-  int hidden_size = static_cast<int>(input_dims[input_dims_size - 1]);
 
-  if (input->Shape() != skip->Shape()) {
+  if (input_dims_size != 3 && input_dims_size != 2) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/skip_layer_norm.cc:64]: nit: input_dims_size check can be moved to right after skip_dims_size check.

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc [hunk 22]
```diff
@@ -60,22 +60,31 @@ Status SkipLayerNorm<T, Simplified>::ComputeInternal(OpKernelContext* ctx) const
...
   }
 
-  const auto& input_dims = input->Shape().GetDims();
-  size_t input_dims_size = input_dims.size();
   if (input_dims_size != 3 && input_dims_size != 2) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                            "input is expected to have 3 or 2 dimensions, got ", input_dims_size);
   }
 
+  if ((input->Shape() != skip->Shape()) && ((skip_dims[0] != 1 || skip_dims_size != 2) && input_dims_size != 3)) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc:78]: The check on CPU and CUDA is the same. We can move the login to a helper function. See example in https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/contrib_ops/cpu/bert/multihead_attention_helper.h.

File: onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc [hunk 23]
```diff
@@ -718,6 +732,55 @@ TEST(SkipLayerNormTest, SkipSimplifiedLayerNormBatch1_Float16) {
...
+          {},
+          epsilon_,
+          batch_size,
+          sequence_length,
+          hidden_size,
+          false,
+          false,
+          false,
+          false,
+          false,
```
Comment (tianleiwu) [inline onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc:778]: Shall add another test case with broadcast_skip = true.

File: .vscode/settings.json [hunk 24]
```diff
@@ -39,5 +39,6 @@
     "cpplint.filters": [
         "-build/include_subdir",
         "-runtime/references"
-    ]
+    ],
+    "cmake.sourceDirectory": "/home/t-khspear/onnxruntime/build/Linux/RelWithDebInfo/_deps/abseil_cpp-src/CMake/install_test_project"
```
Comment (tianleiwu) [inline .vscode/settings.json:43]: revert this change

**Review by tianleiwu — APPROVED** [review]

**Review by wangyems — APPROVED** [review]

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu [hunk 25]
```diff
@@ -127,7 +130,11 @@ __global__ void SkipLayerNormKernelSmall(
   *input_val = *reinterpret_cast<const VecT*>(&input[idx]);
 
   VecT* skip_val = reinterpret_cast<VecT*>(&skip_v);
+  if (skip_broadcasted){
+  *skip_val = *reinterpret_cast<const VecT*>(&skip[idx % skip_size]);
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu:134]: format
